feat(tracer): separate cls from tracer

hekike opened the pull request:
implements: https://github.com/census-instrumentation/opencensus-node/issues/462

**Changes**

- removed the argument notation interface of the `tracer.startChildSpan` because it brings ambiguity into span creation
- Separation of `CoreTracer` and `CoreTracerCls`

**Breaking Changes**

- `startChildSpan` only accepts `options` object
- `Tracer` become `TracerCls`
- `Tracer` is the basic non CLS version now

**Action Items**

- ~~separate tests out to `test-tracer.ts` and `test-tracer-cls.ts`~~ (done)
- ~~Improve non-cls tracer test coverage~~ (done)

**Child of child with CLS Tracer**

It's worth calling out that the CLS implementation still supports only one level of relationships as the `rootSpan` in the namespace context remains the same for child span creation. I guess it's expected based on current instrumentations.

**A couple of further optimization ideas**

- turn `startRootSpan` sync in the non-CLS Tracer
- merge `startRootSpan` and `startChildSpan` in the non CLS Tracer (`startSpan`)
- if we merge the two and make name required, we could get closer to the OpenTracing API: https://github.com/opentracing/opentracing-javascript/blob/master/src/tracer.ts#L79 which I guess after the project convergence is prefered (but maybe this is too early and ambitious)

mayurkale22 commented on packages/opencensus-core/src/trace/model/tracer-cls.ts:
```diff
@@ -0,0 +1,142 @@
+import * as types from './types';
+
+/**
+ * This class represent a tracer with CLS.
```
Nit: `s/represent/represents`

mayurkale22 commented on packages/opencensus-core/src/trace/model/tracer-cls.ts:
```diff
@@ -0,0 +1,142 @@
+
+  /**
+   * Starts a span.
+   * @param [options] span options
```
Nit: `s/@param [options] span options/@param options A SpanOptions object to start a child span./`

mayurkale22 commented on packages/opencensus-core/src/trace/model/tracer.ts:
```diff
@@ -126,43 +109,39 @@ export class CoreTracer implements types.Tracer {
-        parentSpanId = options.spanContext.spanId || '';
-        traceState = options.spanContext.traceState;
+    let traceId;
+    if (options && options.spanContext && options.spanContext.traceId) {
```
Can we just do `options = options || {name: 'span'};` at the start and remove falsiness check on `options` (multiple places below)?

hekike replied:
actually, we could do just:

`function (options = { name: 'span' })`

mayurkale22 commented on packages/opencensus-core/src/trace/model/tracer.ts:
```diff
@@ -229,60 +200,17 @@ export class CoreTracer implements types.Tracer {
    * Starts a span.
-   * @param nameOrOptions Span name string or SpanOptions object.
-   * @param kind Span kind if not using SpanOptions object.
+   * @param [options] span options
```
Same as above

mayurkale22 commented on packages/opencensus-core/src/trace/model/types.ts:
```diff
@@ -516,18 +513,25 @@ export interface Tracer extends SpanEventListener {
+   * @param childOf Span
+   * @param [name] Span name
+   * @param [kind] Span kind
+   * @param [options] Span Options
```
Same as above and just keep options param.

mayurkale22 commented on packages/opencensus-core/src/trace/model/types.ts:
```diff
@@ -516,18 +513,25 @@ export interface Tracer extends SpanEventListener {
+  /** Get and set the currentRootSpan to tracer instance */
+  currentRootSpan: Span;
+
+
```
Nit: Remove extra blank line.

hekike commented:
A couple of further optimization ideas:

- turn `startRootSpan` sync in the non CLS Tracer
- merge `startRootSpan` and `startChildSpan` in the non CLS Tracer (`startSpan`)
- if we merge the two and make name required, we could get closer to the opentracing api: https://github.com/opentracing/opentracing-javascript/blob/master/src/tracer.ts#L79 which I guess after the project convergence is prefered (but maybe this is too early and ambitious)

hekike commented:
It's worth calling out that the CLS implementation still supports only one level of relationships as the `rootSpan` in the namespace context remains the same for child span creation. I guess it's expected based on current instrumentations.

hekike commented:
@mayurkale22 earlier I posted some API ideas as a comment earlier, I just moved them to the PR description now. Could you please share what do you think about them?

mayurkale22 commented:
> @mayurkale22 earlier I posted some API ideas as a comment here, I just moved them to the PR description now. Could you please share what do you think about them?

Thanks for the detailed description, the **Changes** and **Action items** makes sense to me. In terms  of a **breaking changes**, could you please elaborate more on ambiguity on span creation? My initial impression was to keep default tracer as `TracerCls`, so this won't create any breaking change for the 
current users. Also, if feasible, (one approach) we can add `basicTracer` to `Tracing` instance, which users will call to get the basic non CLS version of tracer object. 

> merge startRootSpan and startChildSpan in the non CLS Tracer (startSpan)

SGTM, But would love to see same for CLS Tracer (easier to maintain and switch around). WDYT?

hekike commented:
@mayurkale22 thanks for the feedback! About the API ambiguity: The `startChildSpan`'s original syntax is `startChildSpan(name?: string, kind?: types.SpanKind)`, now we need to add a third `childOf` argument. This is problematic as all the arguments are optional today. If we add it as a first argument, the user needs to pass `undefined` for `tracerCls.startChildSpan(undefined, 'span-name')`. If we add it as the last argument you need to pass `undefined` when you want optional `name` and `kind`. We could build variadic arguments but in my understanding, it is against TypeScript philosophy, but maybe I'm missing something. That's why in think the OpenTracing span creation API is a good tradeoff with a required first string argument and optional secondary options: `tracer.startSpan('foo', { 
childOf: rootSpan })`.

mayurkale22 commented:
> Only keeping the object options notations looks clear and robust solution to me, due to the TypeScript types it's easy to discover the available option properties.

Makes sense to me, please update CHANGELOG.md with a breaking change line. Thanks

kenashcraft commented on packages/opencensus-core/test/test-tracer.ts:
```diff
@@ -354,12 +355,12 @@ describe('Tracer', () => {
-        const child1 =
-            tracer.startChildSpan('child1', types.SpanKind.UNSPECIFIED);
+        const child1 = tracer.startChildSpan(
+            {name: 'child1', kind: types.SpanKind.UNSPECIFIED});
```
To simplify these tests and the API, can `kind` just have a default value?  Then we wouldn't need to specify it again and again.

kenashcraft commented on packages/opencensus-core/src/trace/model/tracer-cls.ts:
```diff
@@ -0,0 +1,142 @@
+import * as types from './types';
+
+/**
+ * This class represent a tracer with CLS.
```
Explain what CLS is and when you would or would not want to use it.

mayurkale22 replied:
+1

hekike commented:
@mayurkale22 @kenashcraft thanks for the review, I addressed the PR comments:

- non-breaking change `CoreTracer` and `CoreTracerBase` separation
- fixes in docs and typos
- documenting CLS usage
- updating `CHANGELOG.md` with `startChildSpan` changes
- cleaning up `startRootSpan` defaults

mayurkale22 commented on packages/opencensus-core/src/trace/model/tracer.ts:
```diff
@@ -14,48 +14,28 @@
 
 /**
- * This class represent a tracer.
+ * This class represents a tracer with Cointunues Local Storage (CLS).
```
Nit: `s/Cointunues/Continuation/`

mayurkale22 commented on examples/http/server.js:
```diff
@@ -42,7 +42,7 @@ function startServer (port) {
 /** A function which handles requests and send response. */
 function handleRequest (request, response) {
-  const span = tracer.startChildSpan('octutorials.handleRequest');
+  const span = tracer.startChildSpan({ name: 'octutorials.handleRequest' });
```
Any update here?

mayurkale22 commented on packages/opencensus-core/src/trace/model/tracer-base.ts:
```diff
@@ -0,0 +1,228 @@
+  private readonly IS_SAMPLED = 0x1;
+  /** A sampler used to make sample decisions */
+  sampler!: samplerTypes.Sampler;
+  /** A configuration for starting the tracer */
```
Please update this comment to something like -> ` An object to log information to`

mayurkale22 commented on packages/opencensus-core/src/trace/model/tracer-base.ts:
```diff
@@ -0,0 +1,228 @@
+        rootSpan.start();
+        return fn(rootSpan);
+      }
+    } else {
```
Can we remove this else, and update the debug msg: `Tracer is inactive, starting new no record root span`?

mayurkale22 commented on packages/opencensus-core/src/trace/model/tracer.ts:
```diff
@@ -14,48 +14,28 @@
- * This class represent a tracer.
+ * This class represents a tracer with Cointunues Local Storage (CLS).
+ *
+ * CLS helps keep tracking the root span over function calls automatically.
```
Can we add ->  `It is capable of storing, propagating and retrieving arbitrary continuation-local data (also called "context").`

hekike replied:
Added

mayurkale22 commented on packages/opencensus-core/test/test-tracer.ts:
```diff
@@ -52,7 +52,7 @@ describe('Tracer', () => {
   });
 
-  /** Should get/set the current RootSpan from tracer instance */
+  // /** Should get/set the current RootSpan from tracer instance */
```
Is this typo or intended stuff?

hekike commented:
@mayurkale22 I addressed your new round of comments.

mayurkale22 commented:
> Action Items (Will do after initial feedback)
separate tests out to test-tracer.ts and test-tracer-cls.ts
Improve non-cls tracer test coverage

@hekike Thanks for the work! What's your plan on above action items. Also, How do you plan to expose basic tracer (with no CLS) using tracing instance? OR as a user, how do I use basic tracer in my app?

hekike commented:
@mayurkale22 I finished the open action items: separates CoreTracer and CoreTracerBase tests and added some context specific tests.

mtwo commented:
@hekike Mayur has been on vacation for the past two weeks, but he's back next Monday. Do you need someone to approve prior to that?

hekike commented:
It would be helpful if we could move this work forward. I'm planning to open further PRs which would require this work.

draffensperger commented on packages/opencensus-core/src/trace/model/tracer-base.ts:
```diff
@@ -0,0 +1,237 @@
+  /** Indicates if the tracer is active */
+  private activeLocal: boolean;
+  /** A configuration for starting the tracer */
+  private config!: configTypes.TracerConfig;
```
Rather than using a `!` would it be possible to assign some reasonable default config?

hekike replied:
The default config leaves in the `opencensus-node` today. I could move it to `opencensus-core` and add as a static property to this class. Then use it here as defaults and consume from here in the `opencensus-node`. My only concern is that this is not a change in this PR and I would prefer to keep the scope focused. But I'm happy to do this as a separate PR.

draffensperger commented on packages/opencensus-core/src/trace/model/tracer-base.ts:
```diff
@@ -0,0 +1,237 @@
+  /** Bit to represent whether trace is sampled or not. */
+  private readonly IS_SAMPLED = 0x1;
+  /** A sampler used to make sample decisions */
+  sampler!: samplerTypes.Sampler;
```
Similar comment here on the sampler - can we assign some reasonable default sampler, e.g. one that samples at the default rate?

draffensperger commented on packages/opencensus-core/src/trace/model/tracer-base.ts:
```diff
@@ -0,0 +1,237 @@
+    // if there is a context propagation, keep the decision
+    if (options && options.spanContext && options.spanContext.options) {
+      propagatedSample =
+          ((options.spanContext.options & this.IS_SAMPLED) !== 0);
```
Nit: can the outer parentheses around this expression be removed?

draffensperger commented on packages/opencensus-core/src/trace/model/tracer-base.ts:
```diff
@@ -0,0 +1,237 @@
+          ((options.spanContext.options & this.IS_SAMPLED) !== 0);
+    }
+
+    let sampleDecision = !!propagatedSample;
```
Could this logic be simplified to `return !!propagatedSample || this.sampler.shouldSample(traceId)`?
